File: 2021/09/14/SubTree.java
/**
 * https://leetcode.com/problems/subtree-of-another-tree/
 * @author Andrew Jarombek
 * @since 9/14/2021
 */

public class SubTree {
    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {}

        TreeNode(int val) { this.val = val; }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        if (root.val == subRoot.val) {
            boolean matches = this.checkMatching(root, subRoot);

            if (matches) {
                return matches;
            }
        }

        boolean matchFound = false;

        if (root.left != null) {
            matchFound = this.isSubtree(root.left, subRoot);
        }

        if (root.right != null) {
            matchFound = matchFound || this.isSubtree(root.right, subRoot);
        }

        return matchFound;
    }

    private boolean checkMatching(TreeNode root, TreeNode subRoot) {
        boolean leftMatches = true;
        boolean rightMatches = true;

        if (root.left != null) {
            if (subRoot.left != null) {
                leftMatches = this.checkMatching(root.left, subRoot.left);
            } else {
                leftMatches = false;
            }
        } else {
            if (subRoot.left != null) {
                leftMatches = false;
            }
        }

        if (root.right != null) {
            if (subRoot.right != null) {
                rightMatches = this.checkMatching(root.right, subRoot.right);
            } else {
                rightMatches = false;
            }
        } else {
            if (subRoot.right != null) {
                rightMatches = false;
            }
        }

        return root.val == subRoot.val && leftMatches && rightMatches;
    }
}

File: 2022/09/13/subtree.js
/**
 * https://leetcode.com/problems/diameter-of-binary-tree/
 * @author Andrew Jarombek
 * @since 9/13/2022
 */

function TreeNode(val, left, right) {
    this.val = (val===undefined ? 0 : val)
    this.left = (left===undefined ? null : left)
    this.right = (right===undefined ? null : right)
}

const isSubtree = function(root, subRoot) {
    let left = false;
    let right = false;

    if (root.left)
        left = isSubtree(root.left, subRoot);
    if (root.right)
        right = isSubtree(root.right, subRoot);

    if (left || right)
        return true;

    if (root.val === subRoot.val)
        return dfs(root, subRoot);

    return false;
};

const dfs = function(root, subRoot) {
    if (root.val !== subRoot.val)
        return false;

    if (!root.left && subRoot.left || !root.right && subRoot.right ||
        !subRoot.left && root.left || !subRoot.right && root.right)
        return false;

    let left = true;
    let right = true;

    if (root.left)
        left = dfs(root.left, subRoot.left)

    if (root.right)
        right = dfs(root.right, subRoot.right)

    return left && right;
}

const tree = new TreeNode(
    3,
    new TreeNode(4, new TreeNode(1), new TreeNode(2)),
    new TreeNode(5)
);
const subTree = new TreeNode(4, new TreeNode(1), new TreeNode(2))
console.assert(isSubtree(tree, subTree));
